### PR TITLE
Added Regex Command (#136)

### LINKED ISSUE #136: Add a Regex Match Checker
Pleezon suggested a feature that allows people to use a slash command that takes a regex pattern and a string and checks whether the string matches the pattern (https://canary.discord.com/channels/648956210850299986/752535909228085348/920221971545989120)

### PR DESCRIPTION
**Pull request by MoonTM-GIT**
Added Regex Command, Closes #136

### CODE REVIEW

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        String string = event.getOption("string").getAsString();
+
+        if (pattern.matcher(string).matches()){
+            return event.replyEmbeds(embedBuilder(true, pattern, string).build());
+        } else {
+            return event.replyEmbeds(embedBuilder(false, pattern, string).build());
+        }
+    }
+
+    private EmbedBuilder embedBuilder(boolean matches, Pattern pattern, String string){
```
Comment (DynxstyGIT) [inline src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java:22]: maybe rename this to something like `buildRegexEmbed` or `buildEmbed` to match the rest of the code
 https://github.com/Java-Discord/JavaBot/blob/e63445021e39b41ce961ef83163f38b7dbc25da4/src/main/java/com/javadiscord/javabot/service/qotw/QOTWJob.java#L50

 https://github.com/Java-Discord/JavaBot/blob/main/src/main/java/com/javadiscord/javabot/commands/staff_commands/Ban.java#L60

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+public class Regex implements SlashCommandHandler {
+
+    @Override
+    public ReplyAction handle(SlashCommandEvent event) throws ResponseException {
+
+        Pattern pattern = Pattern.compile(event.getOption("regex").getAsString());
+        String string = event.getOption("string").getAsString();
+
+        if (pattern.matcher(string).matches()){
+            return event.replyEmbeds(embedBuilder(true, pattern, string).build());
```
Comment (DynxstyGIT) [inline src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java]: you could simplifiy this by doing 
```java
return event.replyEmbeds(embedBuilder(pattern.matcher(string).matches(), pattern, string).build());
```

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        EmbedBuilder eb = new EmbedBuilder()
+                .addField("Regex:", "```" + pattern.toString() + "```", true)
+                .addField("String:", "```" + string + "```", true);
+
+        if (matches) {
+            eb.setTitle("Regex Tester | ✓ Match");
+            eb.setColor(0x30bf56);
+        } else {
+            eb.setTitle("Regex Tester | ✗ No Match");
+            eb.setColor(0xd42f47);
```
Comment (DynxstyGIT) [inline src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java:32]: actually, you could use the slash command colors from the guild config instead

File: src/main/resources/commands.yaml [hunk 4]
```diff
@@ -96,6 +96,12 @@
       required: false
       type: STRING
   handler: com.javadiscord.javabot.commands.user_commands.Poll
+- description: Checks if the given string matches the regex pattern
+  name: regex
+  options:
+    - { description: The regex pattern, name: regex, required: true, type: STRING }
+    - { description: The string, name: string, required: true, type: STRING }
```
Comment (DynxstyGIT) [inline src/main/resources/commands.yaml:103]: please uncollapse this

**Review by DynxstyGIT — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java (hunk 2 above)
Comment (MoonTM-GIT) [reply src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java]: Yup, already did that in a8f411e6364c7d39f582a05c35a4b2ea0b8f68d8

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
+import net.dv8tion.jda.api.requests.restaction.interactions.ReplyAction;
+
+import java.util.regex.Pattern;
+
+public class Regex implements SlashCommandHandler {
+
+    @Override
+    public ReplyAction handle(SlashCommandEvent event) throws ResponseException {
+
```
Comment (DynxstyGIT) [inline src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java:15]: I know that those slash command options are set as 'required', but maybe add tiny checks for those options to make sure they're not null. 

https://github.com/Java-Discord/JavaBot/blob/main/src/main/java/com/javadiscord/javabot/commands/staff_commands/Warn.java#L71-L79

**Review by DynxstyGIT — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                .addField("String:", "```" + string + "```", true);
+
+
+
+        if (matches) {
+            eb.setTitle("Regex Tester | ✓ Match");
+            eb.setColor(config.getSlashCommand().getSuccessColor());
+        } else {
+            eb.setTitle("Regex Tester | ✗ No Match");
+            eb.setColor(config.getSlashCommand().getErrorColor());
```
Comment (DynxstyGIT) [inline src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java:43]: dont initialize the config variable in the `handle` method and just do 
```java
Bot.config.get(guild).getSlashCommand().get{Color};
```
or
```java
var config = Bot.get(guild).getSlashCommand();
... 
eb.setColor(config.getSuccessColor());
...
```

**Review by DynxstyGIT — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+public class Regex implements SlashCommandHandler {
+
+    GuildConfig config;
+
+    @Override
+    public ReplyAction handle(SlashCommandEvent event) throws ResponseException {
+        config = Bot.config.get(event.getGuild());
+
+        Pattern pattern = Pattern.compile(event.getOption("regex").getAsString());
+        String string = event.getOption("string").getAsString();
```
Comment (DynxstyGIT) [inline src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java:23]: You need to check if the Option is null, only after that you can get it as a String

**Review by DynxstyGIT — CHANGES_REQUESTED** [review]

File: src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java (hunk 7 above)
Comment (MoonTM-GIT) [reply src/main/java/com/javadiscord/javabot/commands/user_commands/Regex.java:23]: ah yeah duh

**Review by DynxstyGIT — APPROVED** [review]

**Comment by MoonTM-GIT** [comment]
image

**Pull request merged**